src/engine.rs

```diff
@@ -966,6 +966,90 @@ mod tests {
         assert_eq!(will_force_compact[0], 1);
     }
 
+    #[test]
+    fn test_purge_trigger_force_rewrite() {
+        let dir = tempfile::Builder::new()
+            .prefix("test_purge_trigger_force_write")
+            .tempdir()
+            .unwrap();
+        let cfg = Config {
+            dir: dir.path().to_str().unwrap().to_owned(),
+            target_file_size: ReadableSize::kb(1),
+            purge_threshold: ReadableSize::kb(10),
+            ..Default::default()
+        };
+
+        let engine =
+            RaftLogEngine::open_with_file_system(cfg, Arc::new(ObfuscatedFileSystem::new()))
+                .unwrap();
+        let data = vec![b'x'; 1024];
+        // write 50 small entries into region 1~3, it should trigger force compact
+        // report.
+        for rid in 1..4 {
+            for index in 0..50 {
+                engine.append(rid, index, index + 1, Some(&data[..20]));
+            }
+        }
+        for rid in 4..20 {
+            for index in 0..25 {
+                engine.append(rid, index, index + 1, Some(&data[..50]));
+            }
+        }
+        // write some small entries to trigger purge, some of the should trigger rewrite
+        // but not compact.
+        for rid in 21..100 {
+            engine.append(rid, 1, 2, Some(&data));
+        }
+
+        let check_purge = |pending_regions: Vec<u64>| {
+            let res = engine.purge_expired_files();
+            assert!(res.is_ok());
+            let mut compact_regions = res.unwrap();
+            // sort key in order.
+            compact_regions.sort_unstable();
+            assert_eq!(compact_regions, pending_regions);
+        };
+
+        // purge the 1st time.
+        check_purge(vec![1, 2, 3]);
+        let span = engine.file_span(LogQueue::Append);
+
+        // clean 10 small regions, the file span should not change after purge.
+        for index in 4..8 {
+            engine.compact_to(index, 25);
+        }
+        // compact region 2, then only region 1 still trigger compact.
+        engine.compact_to(3, 30);
+        engine.sync().unwrap();
+        check_purge(vec![1, 2]);
+        let span2 = engine.file_span(LogQueue::Append);
+        assert_eq!(span2, span);
+
+        // purge 3rd.
+        check_purge(vec![1, 2]);
+        let span3 = engine.file_span(LogQueue::Append);
+        assert_eq!(span3, span);
+
+        // pruge the 4th time, it will trigger force rewrite and thus the result should
+        // be empty, and some log files should be purged.
+        check_purge(vec![]);
+        let span4 = engine.file_span(LogQueue::Append);
+        assert!(span4.0 > span.0 + 5);
```

Comment:
This is a little verbose. Could you use a for loop that break when `purge_expired_files` returns empty list, and panic when the loop ends without encountering an empty list.